bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/api/MycroftConnection.java

```diff
@@ -0,0 +1,268 @@
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Establishes and keeps a websocket connection to the Mycroft bus
+ *
+ * @author Gwendal ROULLEAU - Initial contribution. Inspired by the deconz binding.
+ */
+@WebSocket
+@NonNullByDefault
+public class MycroftConnection {
+    private static final AtomicInteger INSTANCE_COUNTER = new AtomicInteger();
+    private final Logger logger = LoggerFactory.getLogger(MycroftConnection.class);
+
+    private final WebSocketClient client;
+    private final String socketName;
+    private final Gson gson;
+
+    private final MycroftConnectionListener connectionListener;
+    private final Map<MessageType, Set<MycroftMessageListener<? extends BaseMessage>>> listeners = new ConcurrentHashMap<>();
+
+    private ConnectionState connectionState = ConnectionState.DISCONNECTED;
+    private @Nullable Session session;
+
+    private static final int TIMEOUT_MILLISECONDS = 3000;
+
+    public MycroftConnection(MycroftConnectionListener listener, WebSocketClient client) {
+        this.connectionListener = listener;
+        this.client = client;
+        this.client.setConnectTimeout(TIMEOUT_MILLISECONDS);
+        this.client.setMaxIdleTimeout(0);
+        this.socketName = "Websocket$" + System.currentTimeMillis() + "-" + INSTANCE_COUNTER.incrementAndGet();
+
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(MessageType.class, new MessageTypeConverter());
+        gson = gsonBuilder.create();
+    }
+
+    public MycroftConnection(MycroftConnectionListener listener) {
+        this.connectionListener = listener;
+        this.client = new WebSocketClient();
+        this.client.setMaxIdleTimeout(0);
+        this.client.setConnectTimeout(TIMEOUT_MILLISECONDS);
+        this.socketName = "Websocket$" + System.currentTimeMillis() + "-" + INSTANCE_COUNTER.incrementAndGet();
+
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(MessageType.class, new MessageTypeConverter());
+        gson = gsonBuilder.create();
+    }
+
+    public void start(String ip, int port) {
+        if (connectionState == ConnectionState.CONNECTED) {
+            return;
+        } else if (connectionState == ConnectionState.CONNECTING) {
+            logger.debug("{} already connecting", socketName);
+            return;
+        } else if (connectionState == ConnectionState.DISCONNECTING) {
+            logger.warn("{} trying to re-connect while still disconnecting", socketName);
+        }
+        Future<Session> futureConnect = null;
+        try {
+            URI destUri = URI.create("ws://" + ip + ":" + port + "/core");
+            client.start();
+            logger.debug("Trying to connect {} to {}", socketName, destUri);
+            futureConnect = client.connect(this, destUri);
+            futureConnect.get(TIMEOUT_MILLISECONDS, TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
```

Comment:
Please avoid catching `Exception` and catch the more specific ones. Because if you catch `Exception` it will also catch runtime exceptions like NPEs and you do not want that. Its probably a `SocketException` or so that will be thrown here.

Please have a look at other places in your code where you catch `Exception` and change them accordingly.

Reply:
Agree, but in fact, jetty client.start() itself is throwing "Exception".
So I don't know what to do except catching Exception ?

Reply:
If that is the case, I guess catching `Exception` is fine as the library (in this case jetty) doesn't give you a choice.